Design docs

benbovy opened the pull request:
I've tried to put together thoughts and ideas found in the issues of this repository and in the sprint doc (https://hackmd.io/UBM5L6YNRlG73e3eVo6vOg) in a design document.

This is still work in progress (although already quite detailed). Feedback and suggestions are very welcome!

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+Examples of common DGGS features that `xdggs` should provide or facilitate:
+
+- convert a DGGS from/to another grid (e.g., a DGGS, a latitude/longitude rectilinear grid, a raster grid, an unstructured mesh)
```
In the discussion on Thursday, we pretty much agreed that this doesn't need to live in `xdggs` (though as part of the work on `xdggs` we need to ensure they exist). `xesmf` and `pyresample` seem like good candidates.

Edit: maybe that's what you meant with the section in the "non-goals"?

benbovy replied:
Yes, what I mean here is that `xdggs` should provide the API for doing these operations (or at least some core API that should facilitate it). `xdggs` does not necessarily need to implement the logic: if there is a well-maintained 3rd party library that does that well, it is preferable to rely on it instead of reinventing the wheel.

About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

benbovy replied:
The 1st sentence of the "goals" section is particularly important: *"unified, high-level and user-friendly API that is deeply integrated with Xarray"*. It is all about the API so I expect that `xdggs` would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

keewis replied:
This is mostly a API design issue: does it make sense to provide a `interpolate` accessor method if we simply forward to another library? Oftentimes, using that library directly is more efficient, since at least in the case of `xesmf` it allows applying the same multiple times without a recomputation of weights. (I don't think there's a universal answer here, it depends on the library and how it works) 

> About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

not sure, but the point is that if they don't we could (try to) extend them instead of implementing this ourselves. Note that those two libraries were examples, if there's a different library that does this already I wouldn't mind using that instead.

keewis replied:
> It is all about the API so I expect that xdggs would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

this would certainly be very convenient, and I didn't think of alternatives, yet (just to have more options to choose from). That will take me some time, though.

benbovy replied:
> This is mostly a API design issue: does it make sense to provide a interpolate accessor method if we simply forward to another library? 

Perhaps not an interpolate accessor but conversion methods in a dggs accessor yes. I think it often makes sense since we're not only regridding but we're also conforming the output to some conventional (or standard?) DGGS representation used by `xdggs` (+ building/setting an index).

It is also about discoverability. We could certainly add tutorials showing how to regrid lat/lon 2D grids to DGGS by directly calling xesmf, pyreseample, etc., and then set the DGGS index. But I find it quite conventional to have some high-level API layer in xdggs to do the regridding and conversion all at once, since it is a very common operation.

Want to create a new DGGS Dataset from another Dataset? Just type `ds.dggs.from_<tab-completion>` and pick up the conversion method that is relevant for your dataset :).

>  at least in the case of xesmf it allows applying the same multiple times without a recomputation of weights. 

I think we could arrange the `xdggs` API so that it allows reusing interpolation weights. That doesn't seem a big issue to me.

> if there's a different library that does this already I wouldn't mind using that instead.

Totally agreed, we are on the same page!

benbovy replied:
What I also tried to state in the doc (non-goals section): *"Likewise, there may be many ways of resampling a grid to a DGGS ; xdggs should support the most common methods but not try to support all of them."*

While I think we should really have conversion methods for converting lat/lon 2D grid to DGGS and vice-versa that handle 80% of the use-cases, it shouldn't prevent reusing directly xesmf, pyresample, etc. directly for the more advanced use cases.

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+`xddgs` should also try to support applications in both GIS and Earth-System communities, which may each use DGGS in slightly different ways (see examples below).
+
+When possible, `xddgs` operations should scale to fine resolutions (millions of cells) leveraging Xarray interoperability with Dask. This might not be always possible, though. Some operations (spatial indexing) may be hard to support at scale and it shouldn't be a high development priority.
```
there's also the idea that whenever necessary (i.e. if the python implementation turns out to be too slow), slow operations need to be written in a lower-level language

benbovy replied:
Yes, for low-level languages / vectorized bindings I also tried to emphasize that in the "goals" and "backends" sections.

Here this sentence is more about horizontal scaling (I can clarify that). While there are some operations that are trivial to support with Dask, other operations like spatial indexing / partitioning / shuffling / sorting are much harder to implement in a distributed fashion. That is not impossible, though, I think we could for example look at [dask-geopandas](https://github.com/geopandas/dask-geopandas). This is pretty hard work, so I wanted to write that it is probably wiser to first try supporting spatial indexing for small amounts of data and then after see how we can scale it horizontally.

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+## Discrete Global Grid Systems
+
+A Discrete Global Grid System (DGGS) can be roughly defined as a partitioning or tesselation of the entire Earth's surface into a finite number of "cells" or "zones". The shape and the properties of these cells generally vary from one DGGS to another. Most DGGSs are also hierarchical, i.e., the cells are aranged on recursively on multiple levels or resolutions. Follow the links in the subsection  ...
```
```suggestion
A Discrete Global Grid System (DGGS) can be roughly defined as a partitioning or tessellation of the entire Earth's surface into a finite number of "cells" or "zones". The shape and the properties of these cells generally vary from one DGGS to another. Most DGGSs are also hierarchical, i.e., the cells are arranged on recursively on multiple levels or resolutions. Follow the links in the subsection below for a more strict and detailed definition of a DGGS.
```

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+DGGSs may be used in various ways, e.g.,
+
+- Applications in Earth-system modelling seem to use DGGS as grids of contiguous, fixed-resolution cells covering the entire Earth's surface or a region of interest (figure 1). This makes easier the analysis of simulation outputs on large extents of the Earth's surface. DGGS may also be used as pyramid data (multiple stacked datasets at different resolutions)
```
```suggestion
- Applications in Earth-system modelling seem to use DGGS as grids of contiguous, fixed-resolution cells covering the entire Earth's surface or a region of interest (figure 1). This makes the analysis of simulation outputs on large extents of the Earth's surface easier. DGGS may also be used as pyramid data (multiple stacked datasets at different resolutions)
```

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+Another draft of DGGS specification can be found here: https://github.com/danlooo/dggs-data-spec.
+
+There are some discrepancies between the proposed standards and popular DGGS libraries (H3, S2, HealPIX). For example regarding the term used to define a grid unit: The two specifications above use "zone", S2/H3 use "cell" and HealPIX uses "pixel". Although in this document we use "cell", the term to choose for `xddgs` is still open for discussion.
```
```suggestion
There are some discrepancies between the proposed standards and popular DGGS libraries (H3, S2, HealPIX). For example regarding the term used to define a grid unit: The two specifications above use "zone", S2/H3 use "cell" and HealPIX uses "pixel". Although in this document we use "cell", the term to choose for `xdggs` is still open for discussion.
```

benbovy replied:
Fixed along with a few other occurrences of that typo.

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+  - plans to switch to pybind11 (no time frame given)
+- [spherely](https://github.com/benbovy/spherely): Python bindings of S2, mostly copying shapely's API
+  - provides numpy-like universal functions
+  - not yest ready for use
```
```suggestion
  - not yet ready for use
```

keewis commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+
+`xdggs.DGGSIndex` is the base class for all Xarray DGGS-aware indexes. It inherits from `xarray.indexes.Index` and has the following specifications:
+
+- It encapsulates an `xarray.indexes.PandasIndex` built from cell ids so that selection and alignment by cell id is possible
```
might need to change that for hierarchical access by cell id

keewis replied:
yep, that's what I meant (hence the edit).

keewis reviewed (commented):
initial feedback with a focus on typos. Anything that requires more discussion we can move to the discussions thread you were planning to create.

benbovy commented on design_doc.md:
```diff
@@ -0,0 +1,268 @@
+Three approaches are possible (non-mutually exclusive):
+
+1. convert cell data into gridded or raster data (choose grid/raster resolution depending on the resolution of the rendered figure) and then reuse existing python plotting libraries (matplotlib, cartopy) maybe through xarray plotting API
+2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API
```
Do you mean hvplot plotting vector data like https://hvplot.holoviz.org/reference/geopandas/polygons.html?

keewis replied:
I meant this: https://discourse.holoviz.org/t/dealing-high-resolution-vector-plots-xarray-holoviz/6310

benbovy replied:
Thanks for the links! Yes it looks like a similar approach to what I suggest by using Datashader with rasterization of DGGS to a dynamic resolution and extent. This could actually be applied for polygon plots: downgrade the DGGS resolution (#18), then aggregate data, then convert it to vector data, then plot it. It would be nice to have some high-level API to do that interactively instead of requiring the user doing all of those steps manually.

benbovy replied:
I rephrased the paragraph below in [5a97e2c](https://github.com/benbovy/xdggs/pull/20/commits/5a97e2c7c405bc8bc9cfdfd801b1cd330046a5ab) to reflect that comment.

keewis commented:
it looks like the remaining points will take a while to discuss, so I don't mind merging this now and continuing the discussion elsewhere.

benbovy commented:
Many thanks @keewis for the feedback. I addressed all your comments that were easily fixable.

I'm +1 for merging this now and continue the discussion in the `Discussions` section of this repository. Follow-up PRs are also welcome to improve the document!

keewis commented:
great, this looks good to me, and let's proceed as you suggested.
